core/controller/src/main/scala/org/apache/openwhisk/core/entitlement/Entitlement.scala

```diff
@@ -231,6 +234,157 @@ protected[core] abstract class EntitlementProvider(
+              } else {
+                Future.successful(())
+              }
+          }
+        }
+        .recoverWith {
+          case t: RejectRequest =>
+            Future.failed(t)
+          case _ =>
+            Future.successful(())
+        }
+    } else if (operation == "remove") {
+      get(entityStore, entityName.toDocId, DocRevision.empty, true)
+        .flatMap { whiskAction =>
+          val currentPermissions = whiskAction.annotations
+            .get(WhiskAction.permissionsFieldName)
+            .map(value => value.convertTo[String])
+            .getOrElse(WhiskAction.defaultPermissions)
+
+          val currentUpdatePermission = currentPermissions.charAt(1)
+          if (currentUpdatePermission == '-') {
+            val errorInfo = s"have no permission to ${operation} this action"
+            Future.failed(RejectRequest(Forbidden, Some(ErrorResponse(errorInfo, transid))))
+          } else {
+            Future.successful(())
+          }
+        }
+        .recoverWith {
+          case t: RejectRequest =>
+            Future.failed(t)
+          case _ =>
+            Future.successful(())
+        }
+    } else if (operation == "invoke") {
+      get(entityStore, entityName.toDocId, DocRevision.empty, true)
+        .flatMap { whiskAction =>
+          val currentPermissions = whiskAction.annotations
+            .get(WhiskAction.permissionsFieldName)
+            .map(value => value.convertTo[String])
+            .getOrElse(WhiskAction.defaultPermissions)
+
+          // the user who is owner by default
+          var currentExecutePermission = currentPermissions.charAt(2)
+          var errorInfo = s"have no permission to ${operation} this action"
+          if (user.namespace.name.asString != entityName.path.root.asString) { // the user who invoke the shared action
+            currentExecutePermission = currentPermissions.charAt(5)
+            errorInfo = s"have no permission to ${operation} this shared action"
+          }
+          if (currentExecutePermission == '-') { //have no permission
+            Future.failed(RejectRequest(Forbidden, Some(ErrorResponse(errorInfo, transid))))
+          } else {
+            Future.successful(())
+          }
+        }
+        .recoverWith {
+          case t: RejectRequest =>
+            Future.failed(t)
+          case _ =>
+            Future.successful(())
+        }
+    } else { // download the code
+      get(entityStore, entityName.toDocId, DocRevision.empty, true)
+        .flatMap { whiskAction =>
+          val currentPermissions = whiskAction.annotations
+            .get(WhiskAction.permissionsFieldName)
+            .map(value => value.convertTo[String])
+            .getOrElse(WhiskAction.defaultPermissions)
+
+          val errorInfo = s"have no permission to download this shared action"
+          val currentDownloadPermission = currentPermissions.charAt(3)
```

Comment:
shouldn't this be `currentPermissions.charAt(0)`? since permission string is like `rwx---` and the first char indicate the "read" permission

Reply:
It seems it is for `the shared user`
![image](https://user-images.githubusercontent.com/11749867/119922929-5f0a0a00-bfa3-11eb-94e8-370e9056f123.png)